### PR TITLE
Emit better warning for overridden recipes

### PR DESCRIPTION
**Pull request by progfolio**
See: #523, #500

Note this change required hoisting the build cache code above the recipe registration code in order to satisfy the byte-compiler.

<details><summary>Test Case</summary>

```emacs-lisp
(straight-bug-report
  :user-dir "better-warning-recipe-override.straight"
  :pre-bootstrap 
  (setq straight-repository-user "progfolio" straight-repository-branch "feat/better-warning-for-recipe-override")
  :post-bootstrap 
  (straight-use-package 'wikinforg)
  (straight-use-package
   '(wikinfo :type git :host github :repo "user/wikinfo")))
```
</details>

- Test run at: `2021-01-12 16:37:26`
- system-type: `gnu/linux`
- straight-version: `prerelease (HEAD -> feat/better-warning-for-recipe-override, fork/feat/better-warning-for-recipe-override) b11b20d 2021-01-12`
- emacs-version: `GNU Emacs 28.0.50 (build 1, x86_64-pc-linux-gnu, X toolkit, cairo version 1.17.4, Xaw3d scroll bars)
 of 2020-12-31`

<details><summary>Output</summary>

```emacs-lisp
Bootstrapping straight.el...
Bootstrapping straight.el...done
Rebuilding all packages due to build cache schema change
Looking for gnu-elpa-mirror recipe → Cloning melpa...
Looking for gnu-elpa-mirror recipe → Cloning melpa...done
Looking for emacsmirror-mirror recipe → Cloning gnu-elpa-mirror...
Looking for emacsmirror-mirror recipe → Cloning gnu-elpa-mirror...done
Looking for emacsmirror-mirror recipe → Cloning el-get...
Looking for emacsmirror-mirror recipe → Cloning el-get...done
Looking for straight recipe → Cloning emacsmirror-mirror...
Looking for straight recipe → Cloning emacsmirror-mirror...done
Building straight...
Building straight...done

Test run with version: prerelease (HEAD -> feat/better-warning-for-recipe-override, origin/feat/better-warning-for-recipe-override) b11b20d 2021-01-12
Cloning wikinforg...
Cloning wikinforg...done
Building wikinforg...
Building wikinforg → Cloning wikinfo...
Building wikinforg → Cloning wikinfo...done
Building wikinforg → Building wikinfo...
Building wikinforg → Building wikinfo...done
Building wikinforg → Cloning org...
Building wikinforg → Cloning org...done
Building wikinforg → Building org...
Building wikinforg → Building org...done
Building wikinforg...
Building wikinforg...done

Warning (straight): "wikinfo" recipe ignored: already installed as a dependency of "wikinforg"

```
</details>

### CODE REVIEW

File: straight.el [hunk 1]
```diff
@@ -3836,6 +3649,197 @@ you ought not to make any changes to it.)"
...
+                                                   straight--build-cache)))
+                           p))
+                    (hash-table-keys straight--build-cache))))
+              ;; Same reasoning as with the previous warning.
+              (straight--warn
+               (concat "Custom recipe given for %S, "
+                       "but it was already pulled in as a dependency of "
+                       "%S")
+               package dependant))
+            (cl-return)))))
```
Comment (progfolio) [inline straight.el:3722]: This is the relevant change

File: straight.el [hunk 2]
```diff
@@ -3836,6 +3649,196 @@ you ought not to make any changes to it.)"
...
+                                   (nth 1 (gethash p
+                                                   straight--build-cache)))
+                           p))
+                    (hash-table-keys straight--build-cache))))
+              ;; Same reasoning as with the previous warning.
+              (straight--warn
+               (concat "Custom recipe for %S ignored because "
+                       "it was installed as a dependency of %S")
+               package dependant))
+            (cl-return)))))
```
Comment (progfolio) [inline straight.el:3721]: relevant change

Comment (progfolio) [inline straight.el:3721]: relevant change

**Comment by raxod502** [comment]
Well, I definitely can't argue with the new message---it's clear and to the point. One thing I will note is that the behavior is not _quite_ how the message describes. We actually do accept the new recipe, and it overrides the old one. However, this probably doesn't have the effect the user was expecting, since the package will have already been cloned and built using the old recipe. Perhaps we should update the recipe resolution system to _actually_ ignore the new recipe. On the other hand, it is very handy to be able to re-evaluate a `straight-use-package` form to update its recipe interactively, then e.g. run `M-x straight-normalize-package` to apply the change. So if we were to try to change the behavior, we should make sure we don't break this kind of use case.

**Comment by progfolio** [comment]
Good points all around. I'll take a look at modifying the recipe resolution to accommodate both use cases.

**Pull request closed without merging**